routing_plugin.go

```diff
@@ -0,0 +1,109 @@
+package loop
```

Comment:
nitty, but this commit is pretty big. Would be happy to pull `routing_plugin.go` out into its own commit with some very rudimental unit tests to reduce size a bit.